File: modules/swagger-codegen/src/test/scala/Java/JavaModelEnumTest.scala
package Java

import io.swagger.codegen.languages.JavaClientCodegen
import io.swagger.models._
import io.swagger.models.properties._
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.{FlatSpec, Matchers}

import scala.collection.JavaConverters._

@RunWith(classOf[JUnitRunner])
class JavaModelEnumTest extends FlatSpec with Matchers {

  it should "convert a java model with an enum" in {
    val enumProperty = new StringProperty()
    enumProperty.setEnum(List("VALUE1", "VALUE2", "VALUE3").asJava)
    val model = new ModelImpl()
      .property("name", enumProperty)

    val codegen = new JavaClientCodegen()
    val cm = codegen.fromModel("sample", model)

    cm.vars.size should be(1)
    val enumVar = cm.vars.get(0)
    enumVar.baseName should be("name")
    enumVar.datatype should be("String")
    enumVar.datatypeWithEnum should be("NameEnum")
    enumVar.name should be("name")
    enumVar.defaultValue should be("null")
    enumVar.baseType should be("String")
    enumVar.isEnum should equal(true)
  }
  
  it should "not override identical parent enums" in {
    
    val identicalEnumProperty = new StringProperty()
    identicalEnumProperty.setEnum(List("VALUE1", "VALUE2", "VALUE3").asJava)
    
    val subEnumProperty = new StringProperty()
    subEnumProperty.setEnum(List("SUB1", "SUB2", "SUB3").asJava)
    
    // Add one enum ptoperty to the parent
    val parentProperties = new java.util.HashMap[String, Property]()
    parentProperties.put("sharedThing", identicalEnumProperty)
    
    // Add TWO enums to the subType model; one of which is identical to the one in parent class
    val subProperties = new java.util.HashMap[String, Property]()
    subProperties.put("sharedThing", identicalEnumProperty)
    subProperties.put("unsharedThing", identicalEnumProperty)

    val parentModel = new ModelImpl();
    parentModel.setProperties(parentProperties);
    parentModel.name("parentModel");
        
    val subModel = new ModelImpl();
    subModel.setProperties(subProperties);
    subModel.name("subModel");
    
    val model = new ComposedModel()
      .parent(new RefModel(parentModel.getName()))
      .child(subModel)
      .interfaces(new java.util.ArrayList[RefModel]())
      
    val codegen = new JavaClientCodegen()
    val allModels = new java.util.HashMap[String, Model]()
    allModels.put(codegen.toModelName(parentModel.getName()), parentModel)
    allModels.put(codegen.toModelName(subModel.getName()), subModel)
    
    val cm = codegen.fromModel("sample", model, allModels)

    cm.name should be("sample")
    cm.classname should be("Sample")
    cm.parent should be("ParentModel")
    cm.imports.asScala should be(Set("ParentModel"))
    
    // Assert that only the unshared/uninherited enum remains
    cm.vars.size should be (1)
    val enumVar = cm.vars.get(0)
    enumVar.baseName should be("unsharedThing")
    enumVar.datatype should be("String")
    enumVar.datatypeWithEnum should be("UnsharedThingEnum")
    enumVar.isEnum should equal(true)
  }
}

File: modules/swagger-codegen/src/test/java/io/swagger/codegen/java/JavaModelEnumTest.java
package io.swagger.codegen.java;

import io.swagger.codegen.CodegenModel;
import io.swagger.codegen.CodegenProperty;
import io.swagger.codegen.DefaultCodegen;
import io.swagger.codegen.languages.JavaClientCodegen;
import io.swagger.oas.models.media.ComposedSchema;
import io.swagger.oas.models.media.Schema;
import io.swagger.oas.models.media.StringSchema;
import org.testng.Assert;
import org.testng.annotations.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import static io.swagger.codegen.CodegenConstants.IS_ENUM_EXT_NAME;
import static io.swagger.codegen.languages.helpers.ExtensionHelper.getBooleanValue;

@SuppressWarnings("static-method")
public class JavaModelEnumTest {

    @Test(description = "convert a java model with an enum")
    public void converterTest() {
        final StringSchema enumSchema = new StringSchema();
        enumSchema.setEnum(Arrays.asList("VALUE1", "VALUE2", "VALUE3"));
        final Schema model = new Schema().type("object").addProperties("name", enumSchema);

        final DefaultCodegen codegen = new JavaClientCodegen();
        final CodegenModel cm = codegen.fromModel("sample", model);

        Assert.assertEquals(cm.vars.size(), 1);

        final CodegenProperty enumVar = cm.vars.get(0);
        Assert.assertEquals(enumVar.baseName, "name");
        Assert.assertEquals(enumVar.datatype, "String");
        Assert.assertEquals(enumVar.datatypeWithEnum, "NameEnum");
        Assert.assertEquals(enumVar.name, "name");
        Assert.assertEquals(enumVar.defaultValue, "null");
        Assert.assertEquals(enumVar.baseType, "String");
        Assert.assertTrue(getBooleanValue(enumVar, IS_ENUM_EXT_NAME));
    }

    @Test(description = "not override identical parent enums")
    public void overrideEnumTest() {
        final StringSchema identicalEnumProperty = new StringSchema();
        identicalEnumProperty.setEnum(Arrays.asList("VALUE1", "VALUE2", "VALUE3"));

        final StringSchema subEnumProperty = new StringSchema();
        subEnumProperty.setEnum(Arrays.asList("SUB1", "SUB2", "SUB3"));

        // Add one enum property to the parent
        final Map<String, Schema> parentProperties = new HashMap<>();
        parentProperties.put("sharedThing", identicalEnumProperty);

        // Add TWO enums to the subType model; one of which is identical to the one in parent class
        final Map<String, Schema> subProperties = new HashMap<>();
        subProperties.put("unsharedThing", subEnumProperty);

        final Schema parentModel = new Schema();
        parentModel.setProperties(parentProperties);
        parentModel.name("parentModel");

        final Schema subModel = new Schema()
                .properties(subProperties)
                .name("subModel")
                .type("object");

        final ComposedSchema composedSchema = new ComposedSchema()
                .addAllOfItem(new Schema().$ref(parentModel.getName()))
                .addAllOfItem(subModel);

        final DefaultCodegen codegen = new JavaClientCodegen();
        final Map<String, Schema> allModels = new HashMap<String, Schema>();
        allModels.put(parentModel.getName(), parentModel);
        allModels.put(composedSchema.getName(), composedSchema);

        final CodegenModel cm = codegen.fromModel("sample", composedSchema, allModels);

        Assert.assertEquals(cm.name, "sample");
        Assert.assertEquals(cm.classname, "Sample");
        Assert.assertEquals(cm.parent, "ParentModel");
        Assert.assertTrue(cm.imports.contains("ParentModel"));

        // Assert that only the unshared/uninherited enum remains
        /** TODO:
        Assert.assertEquals(cm.vars.size(), 1);
        final CodegenProperty enumVar = cm.vars.get(0);
        Assert.assertEquals(enumVar.baseName, "unsharedThing");
        Assert.assertEquals(enumVar.datatype, "String");
        Assert.assertEquals(enumVar.datatypeWithEnum, "UnsharedThingEnum");

         */
    }
}
